man/man6/wmines.6
.TH WMINES 6 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wmines \- minesweeper for W
.SH SYNOPSIS
wmines [-w \fIwidth\fP] [-h \fIheight\fP] [-m \fImines\fP] [-r \fIrounds\fP]
.SH DESCRIPTION
.I Wmines
is a minesweeper game.  Your mission is to isolate all the mines on the
window. A number in a block tells how many mines are in the immediate vicinity
of the block. You need to deduce from the numbers around an un-clicked block
whether it has a mine under it or not. If you think not, click the block. If
the block concealed a mine, you lose, otherwise you'll see a new number (or an
empty block if there are no mines around).
.PP
You can exit the program at the middle of a game with the 'q' key.
.SH ARGUMENTS
.TP
-w
Lets you specify the minefield width in blocks.
.TP
-h
Lets you specify the minefield height in blocks.
.TP
-m
How many mines the minefield will contain.
.TP
-r
How many rounds you want to play (with increasing amount of mines, losing game
quits the program).
.PP
Minefield width, height and amount of mines have reasonable limits. Minimum
values are:
.PP
wmines -w 10 -h 10 -m 10 -r 1
.SH LAST CHANGED
11/1995 by Eero Tamminen


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


games/wmines.c
/*
 * wmines.c, a part of the W Window System
 *
 * Copyright (C) 1995 by Eero Tamminen
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- minesweeper game clone for W
 */

#include <stdio.h>		/* printf() */
#include <stdlib.h>		/* rand(), srand() */
#include <time.h>		/* time() */
#include <Wlib.h>

/* W specific stuff */

/* space lost for window borders horizontally and vertically */
#define HOR_BORDERS	8
#define VER_BORDERS	18
#define WIN_PROPERTIES	(W_MOVE | W_TITLE | W_CLOSE | EV_MOUSE | EV_KEYS)

#ifdef WTINYSCREEN
  #define BLOCK_W	10			/* block size */
  #define BLOCK_H	10
  #define WIDTH 	15			/* default field size */
  #define HEIGHT	14
  #define MIN_W		6			/* minimum field size */
  #define MIN_H		6
#else
  #define BLOCK_W	12			/* block size */
  #define BLOCK_H	12
  #define WIDTH 	20			/* default field size */
  #define HEIGHT	20
  #define MIN_W		10			/* minimum field size */
  #define MIN_H		10
#endif
#define MINES	(WIDTH * HEIGHT / 10)
#define ROUNDS	1

/* font settings */

#define FNAME	"fixed"
#ifdef WTINYSCREEN
  #define FSIZE (BLOCK_W - 2)
#else
  #define FSIZE	(BLOCK_W - 4)		/* smaller than blocksize */
#endif

/* global variables */
static WWIN *Window;
static WFONT *Font;

typedef struct
{
  unsigned char **field;		/* (height + 2) * (width + 2) */
  int block_w;
  int block_h;
  int width;
  int height;
  int mines;
  int rounds;
} MINEFIELD;

MINEFIELD MineField =
{
  0, BLOCK_W, BLOCK_H, WIDTH, HEIGHT, MINES, ROUNDS
};

/* bit that indicates that block hasn't been yet 'turned' */
#define HIDDEN	64


/* prototypes */
static void process_args(char *argv[]);
static int  initialize(char *argv[]);
static void exit_game(int signal);
static void random_field(void);
static void first_clear(short *todo);
static void seed_clear(short *todo, short x, short y);
static void number_block(short num, short x, short y);
static void draw_game(void);
static void show_mines(void);
static void wait_input(void);

int main(int argc, char *argv[])
{
  short mx, my, todo, hit;

  /* check command line arguments */
  process_args(argv);

  /* initialize W and Minefield struct */
  if(!initialize(argv))
    exit_game(1);

  srand(time(0));

  do
  {
    w_settitle(Window, " MINES ");
    draw_game();
    random_field();
    todo = MineField.width * MineField.height - MineField.mines;
    w_setmode(Window, M_CLEAR);
    first_clear(&todo);

    hit = 0;
    do
    {
      wait_input();
      w_querymousepos(Window, &mx, &my);
      mx = mx / MineField.block_w + 1;
      my = my / MineField.block_h + 1;
      if(MineField.field[mx][my] >= HIDDEN) {
	if(MineField.field[mx][my] > (HIDDEN + 8))
	  hit = 1;
	else
	  seed_clear(&todo, mx, my);
      }
    }
    while(!hit && todo);

    show_mines();
    if(todo)
      w_settitle(Window, " KABOOM! ");
    else
      w_settitle(Window, " SURVIVED... ");
    wait_input();
    MineField.mines += MineField.mines / 4;

  } while(--MineField.rounds && !todo);

  exit_game(0);
  return 0;
}

/* process command line arguments */
static void process_args(char *argv[])
{
  short err = 0, opt, arg;

  while(*++argv)
  {
    /* option, letter (only one), with an argument? */
    if(**argv == '-' && (opt = *(*argv+1)) && !*(*argv+2) && *(argv+1))
    {
      arg = atoi(*++argv);

      switch(opt)
      {
	case 'w':
	  if(arg >= MIN_W)
	    MineField.width = arg;
	  break;

        case 'h':
	  if(arg >= MIN_H)
	    MineField.height = arg;
	  break;

        case 'r':
	  if(arg > ROUNDS)
	    MineField.rounds = arg;
	  break;

        case 'm':
          MineField.mines = arg;
	  break;

	default:
	  err = 1;
      }
    }
    else
      err = 1;
  }

  /* not too easy nor hard... */
  if(MineField.mines < (arg = MineField.width * MineField.height / 10) ||
     MineField.mines >  arg * 4)
    MineField.mines = arg;

  if (err)
  {
    fprintf(stderr, "usage: wmines [-w <width>] [-h <height>] [-m <mines>] [-r <rounds>]\n");
    exit(1);
  }
}

/* Initialize W stuff (connect server, open window, set font/title)
 * and allocate Minefield. Return 0 on error.
 */
static int initialize(char *argv[])
{
  WSERVER *wserver;
  short w_width, w_height, i;

  if (!(wserver = w_init()))
  {
    fprintf(stderr, "%s: can't connect to wserver\n", *argv);
    return 0;
  }
  w_width = wserver->width;
  w_height = wserver->height;

  /* limit window size to screen size */
  w_width  -= HOR_BORDERS;
  w_height -= VER_BORDERS;
  if(MineField.block_w * MineField.width > w_width)
    MineField.width  = w_width / MineField.block_w;
  if(MineField.block_h * MineField.height > w_height)
    MineField.height = w_height / MineField.block_h;

  /* calculate window size */
  w_width  = MineField.block_w * MineField.width;
  w_height = MineField.block_h * MineField.height;

  if(!(Window = w_create(w_width, w_height, WIN_PROPERTIES)))
  {
    fprintf(stderr, "%s: unable to create a window\n", *argv);
    return 0;
  }
  if (!(Font = w_loadfont(FNAME, FSIZE, 0)))
  {
    fprintf(stderr, "%s: unable to load '" FNAME "%d' font\n", *argv, FSIZE);
    return 0;
  }
  w_setfont(Window, Font);
  if(w_open(Window, UNDEF, UNDEF) < 0)
  {
    fprintf(stderr, "%s: unable to open a window\n", *argv);
    return 0;
  }

  /* allocate space for the field pointers and rows */
  MineField.field = (unsigned char **)malloc(sizeof(void *) * (MineField.width + 2));
  for(i = 0; i < MineField.width + 2; i ++)
    if(!(MineField.field[i] = (unsigned char *)malloc((MineField.height + 2))))
    {
      fprintf(stderr, "%s: unable to allocate minefield\n", *argv);
      return 0;
    }
  return 1;
}

/* signal / exit handler */
static void exit_game(int signal)
{
  int i;

  if(MineField.field)
  {
    for(i = 0; i < MineField.width + 2; i++)
      free(MineField.field[i]);
    free(MineField.field);
  }
  if(Window)
  {
    w_close(Window);
    w_delete(Window);
  }
  exit(signal);
}


/* randomize the minefield, check how many mines any square adjoins
 * > 8 means that the square contains a mine itself
 */
static void random_field(void)
{
  int i, x, y;

  for(y = 0; y < MineField.height + 2; y++)
    for(x = 0; x < MineField.width + 2; x++)
      MineField.field[x][y] = HIDDEN;

  for(i = 0; i < MineField.mines; i++)
  {
    /* search an empty place */
    do
    {
      x = rand() % MineField.width  + 1;
      y = rand() % MineField.height + 1;
    } while(MineField.field[x][y] > (HIDDEN + 8));

    /* position mine */
    MineField.field[x][y] += 9;

    /* increase mine count in the neighbourhood */
    MineField.field[x-1][y] += 1;
    MineField.field[x+1][y] += 1;
    y -= 1;
    MineField.field[x][y]   += 1;
    MineField.field[x-1][y] += 1;
    MineField.field[x+1][y] += 1;
    y += 2;
    MineField.field[x][y]   += 1;
    MineField.field[x-1][y] += 1;
    MineField.field[x+1][y] += 1;
  }
}

/* makes one clearing on start of the game */
static void first_clear(short *todo)
{
  short x, y, i = MineField.width * MineField.height;

  /* get an empty place */
  do
  {
    x = rand() % MineField.width  + 1;
    y = rand() % MineField.height + 1;
  } while(i-- && MineField.field[x][y] > HIDDEN);

  /* and clear it */
  seed_clear(todo, x, y);
}


/* clear all the empty blocks around (x,y) using recursive 'seed-fill' */
static void seed_clear(short *todo, short x, short y)
{
#define STACK	40
#define PUSH(a,b) {if(idx < STACK-1) { stack[idx++] = a; stack[idx++] = b; }}
  short stack[STACK], idx = 0, start, above, below;
  short num, width = MineField.width;
  unsigned char **field = MineField.field;

  for(;;)
  {
    /* don't start if not an empty block */
    if((num = field[x][y] & ~HIDDEN))
    {
      number_block(num, x, y);
      (*todo)--;
      return;
    }

    /* search left edge (hidden or visible) */
    while(x > 1 && !(field[x][y] & ~HIDDEN))
      x--;

    start = above = below = x;
    do
    {
      /* check above */
      if(--y > 0 && (num = field[x][y]) & HIDDEN)
      {
	num &= ~HIDDEN;
	if(num)
	{
	  above = x;
	  number_block(num, x, y);
	  (*todo)--;
	}
	else
	  if(above <= x)
	  {
	    PUSH(x, y);
	    above = width;
	  }
      }
      y++;

      /* check below */
      if(++y <= MineField.height && (num = field[x][y]) & HIDDEN)
      {
	num &= ~HIDDEN;
	if(num)
	{
	  below = x;
	  number_block(num, x, y);
	  (*todo)--;
	}
	else
	  if(below <= x)
	  {
	    PUSH(x, y);
	    below = width;
	  }
      }
      y--;

      /* check current */
      num = field[x][y];
      if(num & HIDDEN)
      {
        num &= ~HIDDEN;
	number_block(num, x, y);
	(*todo)--;
      }
    /* until right edge */
    } while((x == start || !num) && x++ < width);

    /* while something in stack, POP(x,y) */
    if(idx > 1)
    {
      /* last-in, first-out! */
      y = stack[--idx];
      x = stack[--idx];
    }
    else
      return;
  }
}

/* clear an area and put a number on it */
static void number_block(short number, short x, short y)
{
  short block_w = MineField.block_w;
  short block_h = MineField.block_h;

  MineField.field[x--][y--] = number;	/* clear hidden bit */

  x *= block_w;
  y *= block_h;

  /* clear (set in main) area */
  w_pbox(Window, x, y, block_w - 1, block_h - 1);

  if(!number)
    return;

  /* output number centered */
  x += (block_w - Font->widths[number + '0']) / 2;
  y += (block_h - Font->height) / 2;
  w_printchar(Window, x, y, number + '0');
}


/* draws the initial game window contents */
static void draw_game(void)
{
  short block_w = MineField.block_w;
  short block_h = MineField.block_h;
  short win_w = MineField.width  * block_w;
  short win_h = MineField.height * block_h;
  short x, y;

  w_dpbox(Window, 0, 0, win_w, win_h);

  w_setmode(Window, M_CLEAR);
  for(y = 0; y < win_h; y += block_h)
    for(x = 0; x < win_w; x += block_w)
    {
      w_hline(Window, x, y, x + block_w - 2);
      w_vline(Window, x, y + 1, y + block_h - 2);
    }

  w_setmode(Window, M_DRAW);
  for(y = 0; y < win_h; y += block_h)
    for(x = 0; x < win_w; x += block_w)
    {
      w_hline(Window, x + 1, y + block_h - 2, x + block_w - 2);
      w_vline(Window, x + block_w - 2, y + 1, y + block_h - 3);
    }
  for(x = -1; (x += block_w) < win_w;)
    w_vline(Window, x, 0, win_h - 1);
  for(y = -1; (y += block_h) < win_h;)
    w_hline(Window, 0, y, win_w - 1);
}

/* show all the mines */
static void show_mines(void)
{
  short block_w = MineField.block_w;
  short block_h = MineField.block_h;
  short x, y;

  w_setmode(Window, M_DRAW);
  for(y = 1; y <= MineField.height; y++)
    for(x = 1; x <= MineField.width; x++)
      if(MineField.field[x][y] > (HIDDEN + 8))
        w_pbox(Window,
	  (x - 1) * block_w + 2, (y - 1) * block_h + 2,
	  block_w-5, block_h-5);
}

static void wait_input(void)
{
  WEVENT *ev;

  /* until mouse button released or exit */
  for(;;)
  {
    while(!(ev = w_queryevent(NULL, NULL, NULL, -1)));
    switch(ev->type)
    {
      case EVENT_MRELEASE:
        return;

      case EVENT_KEY:
        if((ev->key & 0xff) == 'q')
	  exit_game(0);
	break;

      case EVENT_GADGET:
        if(ev->key == GADGET_EXIT || ev->key == GADGET_CLOSE)
	  exit_game(0);
    }
  }
}
